File: packages/core/src/modules/hassault/hitbox.ts
/**
 * The body a shot is resolved against, client side.
 *
 * The mirror of `backend/modules/hassault/hitbox.py`, and deliberately shaped the
 * same way: the server is the authority, this is what the prediction runs against,
 * and the two must not be able to disagree about how tall a player is.
 *
 * **Injected, never fetched.** `world.ts` and `player.ts` are three-free and
 * import-free on purpose so the conformance suite can replay
 * `__tests__/physics-vectors.json` headless against the same code the game runs.
 * A module that called `fetch` here would end that, so the session pushes the
 * served spec in with `setHitbox` at join time and this file stays pure.
 *
 * The exported `DEFAULT_HITBOX` is the shipped body — what the vectors were
 * generated against, and what prediction uses until the server says otherwise.
 */

/** One body, in cubes. Field-for-field the server's `HitboxSpec.to_dict()`. */
export interface HitboxSpec {
  /** Content hash of the hit-deciding dimensions. Shown in the tuning lab, and
   * stamped into `physics-vectors.json` so a stale fixture cannot pass. */
  specId: string;
  /** Only `cylinder` exists today. Carried so a client can refuse a shape it does
   * not know how to draw, rather than drawing the wrong one silently. */
  shape: string;
  radius: number;
  eyeHeight: number;
  aboveEye: number;
  standingHeight: number;
  crouchEyeScale: number;
  crouchEyeHeight: number;
  crouchHeight: number;
  /** Crouched height as a fraction of standing — what an avatar is squashed by. */
  crouchScale: number;
  /** Top band of the body that takes the weapon's head multiplier. */
  headBand: number;
  fitTolerance: number;
  eyeTolerance: number;
}

/**
 * The shipped body: AssaultCube's `entity.h` defaults.
 *
 * Derived values are spelled out rather than computed, because the server serves
 * them computed — two implementations of `crouchHeight` is two chances to round it
 * differently, which is precisely the class of drift this module exists to close.
 */
export const DEFAULT_HITBOX: HitboxSpec = {
  specId: '86d9f2779917',
  shape: 'cylinder',
  radius: 1.1,
  eyeHeight: 4.5,
  aboveEye: 0.7,
  standingHeight: 5.2,
  crouchEyeScale: 0.75,
  crouchEyeHeight: 3.375,
  crouchHeight: 4.075,
  crouchScale: 4.075 / 5.2,
  headBand: 1.0,
  fitTolerance: 0.35,
  eyeTolerance: 0.15,
};

let active: HitboxSpec = DEFAULT_HITBOX;

/** The spec in force. Read per call, never captured in a default argument — a
 * default binds at module load, so a tuned body would reach everything except the
 * code that decides the hit. */
export function currentHitbox(): HitboxSpec {
  return active;
}

/**
 * Adopt the served spec.
 *
 * Called once when a session starts, and again when the tuning lab changes the
 * body. A `null` restores the shipped one, which is what a failed fetch should do:
 * predicting against the default is wrong only if somebody has tuned it, whereas
 * predicting against nothing does not work at all.
 */
export function setHitbox(spec: HitboxSpec | null): HitboxSpec {
  active = spec ?? DEFAULT_HITBOX;
  return active;
}

/** Body height mid-crouch, `crouch` being the 0..1 animation fraction. */
export function heightAt(crouch: number, spec: HitboxSpec = active): number {
  return spec.standingHeight + (spec.crouchHeight - spec.standingHeight) * crouch;
}

/** Eye height mid-crouch — where the camera sits and where a shot leaves from. */
export function eyeAt(crouch: number, spec: HitboxSpec = active): number {
  return spec.eyeHeight + (spec.crouchEyeHeight - spec.eyeHeight) * crouch;
}

File: backend/modules/hassault/hitbox.py
"""The body a shot is resolved against — served, versioned, and deliberately not final.

Every number in here used to be a constant in `physics.py` copied by hand into
`world.ts`, `player.ts`, `avatars.ts` and `bodies.rs`. Four copies of a figure
nobody has finished choosing: there is no wall penetration yet, the head band has
never been checked against a real character model, and the relationship between
what a body *looks* like and what it *is* has only ever been asserted. So this
module exists to make changing the hitbox cheap and observable rather than
frightening, and the clients read it instead of holding their own copy — the
`plane_order` / `zoom_levels` precedent.

**The spec is the authority; a mesh never is.** An avatar is fitted to this, not
the other way round, and no renderer gets to widen a body by drawing it wide. That
much is fixed even though the numbers are not.

## The version is a hash, not a label

`spec_id` is derived from the numbers themselves. That is the whole point: a
hand-maintained revision is a thing you forget to bump, and the failure mode of
forgetting is the worst one available here — the Python simulation and the two
client implementations replaying a fixture that no longer describes any of them,
agreeing with each other about a body that does not exist. A content hash cannot
be forgotten. Change a number, the id changes, and
`physics-vectors.json` fails until it is regenerated deliberately.

## What is *not* here

- **Step height, speeds, gravity, friction.** Those are movement, not the body.
  They stay in `physics.py`.
- **Wall penetration.** A bullet stopping at a wall is a property of the *wall*
  (`weapons.py` + the world's materials), and the temptation to compensate for a
  missing penetration model by widening a body is exactly what this module exists
  to make visible.
"""

from __future__ import annotations

import hashlib
import json
from dataclasses import dataclass, replace
from typing import Any


@dataclass(frozen=True, slots=True)
class HitboxSpec:
    """One body, in cubes.

    Only `cylinder` exists today. `shape` is carried anyway so that growing a
    capsule, or a segment list finer than the current head/body split, is a value
    change rather than a schema change — and so a client can refuse a shape it does
    not know how to draw instead of drawing the wrong one silently.
    """

    #: Radius of the collision cylinder. Also what the movement code keeps clear,
    #: which is why a body is 2.2 cubes wide and needs 3 cells of clearance.
    radius: float
    #: Standing eye height: where the camera sits and where a shot leaves from.
    eye_height: float
    #: How much body there is above the eye. Standing height is the sum.
    above_eye: float
    #: Crouching multiplies the *eye* height by this; `above_eye` is unchanged, so
    #: a crouched body is shorter by `eye_height * (1 - crouch_eye_scale)`.
    crouch_eye_scale: float
    #: The top band of the body that takes the weapon's head multiplier. A band
    #: measured down from the top rather than an absolute height, because a band
    #: pinned to a standing figure sits above a crouched one entirely and makes
    #: crouching an accidental headshot immunity.
    head_band: float
    #: How far outside the cylinder an avatar mesh may reach before its fit report
    #: is a failure. Non-zero on purpose: cloth, packs and a rifle silhouette are
    #: allowed to exceed the body precisely because the server never asks the mesh
    #: anything. Tightening this is how you make avatars honest; it lives here so
    #: it is tuned alongside the body it is a tolerance on.
    fit_tolerance: float = 0.35
    #: How far the rig's eye bone may sit from `eye_height`. Tighter than
    #: `fit_tolerance` because the eye is not decoration — a first-person camera
    #: and a third-person head are supposed to be the same place.
    eye_tolerance: float = 0.15
    shape: str = "cylinder"

    # -- derived ---------------------------------------------------------------

    @property
    def standing_height(self) -> float:
        return self.eye_height + self.above_eye

    @property
    def crouch_eye_height(self) -> float:
        return self.eye_height * self.crouch_eye_scale

    @property
    def crouch_height(self) -> float:
        return self.crouch_eye_height + self.above_eye

    @property
    def crouch_scale(self) -> float:
        """Crouched height as a fraction of standing — what an avatar is squashed
        by. Derived rather than stored so it cannot disagree with the heights."""
        return self.crouch_height / self.standing_height

    def height_at(self, crouch: float) -> float:
        """Body height mid-crouch, `crouch` being the 0..1 animation fraction."""
        return (
            self.standing_height + (self.crouch_height - self.standing_height) * crouch
        )

    def eye_at(self, crouch: float) -> float:
        """Eye height mid-crouch."""
        return self.eye_height + (self.crouch_eye_height - self.eye_height) * crouch

    # -- identity --------------------------------------------------------------

    def _canonical(self) -> str:
        """The numbers that decide a hit, in a stable textual form.

        `fit_tolerance` and `eye_tolerance` are **excluded**: they govern whether an
        art asset is acceptable, not where a bullet lands, so tightening them must
        not invalidate a physics fixture that is still perfectly accurate.
        """
        return json.dumps(
            {
                "shape": self.shape,
                "radius": self.radius,
                "eye_height": self.eye_height,
                "above_eye": self.above_eye,
                "crouch_eye_scale": self.crouch_eye_scale,
                "head_band": self.head_band,
            },
            sort_keys=True,
            separators=(",", ":"),
        )

    @property
    def spec_id(self) -> str:
        """Content hash of the hit-deciding numbers. Stamped into
        `physics-vectors.json`; a fixture carrying a different one is stale."""
        return hashlib.sha256(self._canonical().encode()).hexdigest()[:12]

    def to_dict(self) -> dict[str, Any]:
        """The shape both clients read. Derived values are included rather than
        left for each client to recompute — three implementations of
        `crouch_height` is three chances to round it differently."""
        return {
            "specId": self.spec_id,
            "shape": self.shape,
            "radius": self.radius,
            "eyeHeight": self.eye_height,
            "aboveEye": self.above_eye,
            "standingHeight": self.standing_height,
            "crouchEyeScale": self.crouch_eye_scale,
            "crouchEyeHeight": self.crouch_eye_height,
            "crouchHeight": self.crouch_height,
            "crouchScale": self.crouch_scale,
            "headBand": self.head_band,
            "fitTolerance": self.fit_tolerance,
            "eyeTolerance": self.eye_tolerance,
        }


#: The shipped body. AssaultCube's `entity.h` defaults, which is where the numbers
#: came from and not an argument that they are right for this game.
DEFAULT = HitboxSpec(
    radius=1.1,
    eye_height=4.5,
    above_eye=0.7,
    crouch_eye_scale=0.75,
    head_band=1.0,
)

_override: HitboxSpec | None = None


def current() -> HitboxSpec:
    """The spec in force. Process-global for the same reason the karaoke session
    is: the match server, the REST surface and the agent tools must not be able to
    disagree about how tall a player is, and none of them owns the others."""
    return _override or DEFAULT


def set_override(spec: HitboxSpec | None) -> HitboxSpec:
    """Replace the live spec (the tuning lab) or clear it with `None`.

    Deliberately **not** a setting. `SettingValue` is a scalar and
    `GET /api/settings` hands the whole bag to every plugin; more to the point a
    hitbox is one coherent object, and a half-applied body — a new radius against
    an old head band — is a state no code here should ever have to consider.
    """
    global _override
    _override = spec
    return current()


def tune(**changes: float) -> HitboxSpec:
    """Override a few fields of the current spec, keeping the rest."""
    return set_override(replace(current(), **changes))


def reset() -> HitboxSpec:
    """Back to the shipped body."""
    return set_override(None)
